anta/tests/interfaces.py

```diff
@@ -778,3 +779,98 @@ def test(self) -> None:
+    - If speed with lane (100g-8) then verify mode as full duplex and speed/lanes as per input.
+
+    Expected Results
+    ----------------
+    * Success: The test will pass if an interface is configured with the correct speed, lanes, auto-negotiation and mode as full duplex.
+    * Failure: The test will fail if an interface is not found, the speed or lanes do not match with input,
+               auto-negotiation is not correct or mode is not full duplex.
+
+    Examples
+    --------
+    ```yaml
+    anta.tests.interfaces:
+      - VerifyInterfacesSpeed:
+          interfaces:
+            - name: Ethernet2
+              speed: 100g
+            - name: Eth3
+              speed: 1-0
+            - name: Eth3
+              speed: auto 100-1
+            - name: Eth2
+              speed: auto 2.5g
+            - name: Et3
+              speed: forced 1
+    ```
+    """
+
+    name = "VerifyInterfacesSpeed"
+    description = "Verifies the speed, lanes, auto-negotiation status, and mode as full duplex for interfaces."
+    categories: ClassVar[list[str]] = ["interfaces"]
+    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [AntaCommand(command="show interfaces")]
+
+    class Input(AntaTest.Input):
+        """Inputs for the VerifyInterfacesSpeed test."""
+
+        interfaces: list[InterfaceDetail]
+        """List of interfaces to be tested"""
+
+        class InterfaceDetail(BaseModel):
+            """Detail of an interface."""
+
+            name: EthernetInterface
+            """Name of the interface."""
+            speed: InterfaceSpeedPattern
+            """Speed of an interface in Gigabits per second."""
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """Main test function for VerifyInterfacesSpeed."""
+        self.result.is_success()
+        command_output = self.instance_commands[0].json_output
+
+        # Iterate over all the interfaces
+        for interface in self.inputs.interfaces:
+            intf = interface.name
+
+            # Check if interface exists
+            if not (interface_output := get_value(command_output, f"interfaces.{intf}")):
+                self.result.is_failure(f"Interface `{intf}` not found.")
+                continue
+
+            auto_negotiation = interface_output.get("autoNegotiate")
+            duplex = interface_output.get("duplex")
+            actual_speed = interface_output.get("bandwidth")
+            actual_lanes = interface_output.get("lanes")
+            speed, lanes = extract_speed_and_lane(interface.speed)
+
+            # Collecting actual interface details
+            actual_interface_output = {
+                "auto negotiation": auto_negotiation if "auto" in interface.speed else None,
+                "duplex mode": duplex,
+                "speed": f"{custom_division(actual_speed, 1000000000)}Gbps" if interface.speed != "auto" else None,
```

Comment:
so bandwith is in bps, why don't we simply convert the input string in bps after parsing as an Integer rather than using some custom division ? Is it for the error message?

Reply:
Yes, for failure we are providing the details in Gbps so even if we compare bps rate then for failure we have to convert it.

Reply:
I understand that but here we are
1. parsing an input, getting a string and appending Gbps
2. converting an output field and appending Gbps to it after custom division
3. compare 2 values
4. Maybe generate an error

Why not simply do:
1. parse input and convert it to bps
2. compare 2 values
3. format error message if required (maybe with division)

?

Here we are artifically hardcoding Gbps in our test